schema-layer/data-structures/hashmap.md

```diff
@@ -0,0 +1,292 @@
+* `bitWidth`: The number of bits to use at each level of the data structure for determining the index of the entry or a link to the next level of the data structure to continue searching. The equation `2`<sup>`bitWidth`</sup> yields the arity of the HashMap nodes, i.e. the number of storage locations for buckets and/or links to child nodes.
+* `bucketSize`: The maximum array size of entry storage buckets such that exceeding `bucketSize` causes the creation of a new child node to replace entry storage.
+
+### Node properties
+
+Each node in a HashMap data structure contains:
+
+* `data`: An array, with a length of one to `2`<sup>`bitWidth`</sup>.
+* `map`: A bitfield where the first `2`<sup>`bitWidth`</sup> bits are used to indicate whether a bucket or child node link is present at each possible index of the node.
+
+An important property of a HAMT is that the `data` array only contains active elements. Indexes in a node that do not contain any values (in buckets or links to child nodes) are not stored and the `map` bitfield is used to determine the `data` whether values are present and the array index of present values using a [`popcount()`](https://en.wikipedia.org/wiki/Hamming_weight). This allows us to store a maximally compacted `data` array for each node.
+
+### Schema
+
+The **root block** of an IPLD HashMap contains the same properties as all other blocks, in addition to configuration data that dictates how the algorithm below traverses and mutates the data structure.
+
+See [IPLD Schemas](/schema-layer/schemas) for a definition of this format.
+
+```ipldsch
+# Root node layout
+type HashMapRoot struct {
+  hashAlg String
+  bitWidth Int
+  bucketSize Int
+  map Bytes
+  data [ Element ]
+}
+
+# Non-root node layout
+type HashMapNode struct {
+  map Bytes
+  data [ Element ]
+}
+
+type Element union {
+  | Link "link"
+  | Bucket "bucket"
+} representation keyed
+
+type Bucket list [ BucketEntry ]
+
+type BucketEntry struct {
+  key Bytes
+  value Value
+} representation tuple
+
+type Value union {
+  | Bool bool
+  | String string
+  | Bytes bytes
+  | Int int
+  | Float float
+  | Map map
+  | List list
+  | Link link
+} representation kinded
+```
+
+Notes:
+
+* `hashAlg` in the root block is a string name that corresponds to a [multihash](https://github.com/multiformats/multihash) as found in the [multiformats table](https://github.com/multiformats/multicodec/blob/master/table.csv).
+* `bitWidth` in the root block must be at least `2` and no more than `12`. ***(TODO: this is arbitrary and currently IAMap only supports a maximum of `8` and go-hamt-ipld _only_ supports `8`. `12` yields a maximum `data` array length of `4096`, is it reasonable to go higher? A two-byte maximum would be neat, but `65,536` elements is maybe too much.)***
+* `bucketSize` in the root block must be at least `1`. ***(TODO: does it need a maximum?)***
```

Comment:
👍 but the question of well-defined interop space may be important here too, see above about `bitWidth`. Do you want people showing up in the Go repo saying "I made this HashMap in my browser but the Go implementation won't read it? Being able to fall back to saying "you pushed `bucketSize` beyond `<well defined point>` and you hit caveat emptor space, sorry" might be nice. Or do we just let those points evolve and each implementation document their limits?

Reply:
You're right, we should definitely recommend a maximum. I just don't want implementations _rejecting_ or failing to handle HAMTs with large buckets (even if the block size is reasonable). (i.e., SHOULD versus MUST)

Reply:
Here's what I've done for both `bitWidth` and `bucketSize`. In this section I've just specified the minimums ("should" for `bitWidth`, "must" for `bucketSize` because `<=0` doesn't make sense). Down below in the "Implementation Defaults" section I've added this type of thing to them both: "No maximum is specified, however implementers should be aware that interoperability problems may arise with very large `bucketSize` values."